src/docs/05-go-client/19-workflow-non-deterministic-error.md

```diff
@@ -0,0 +1,178 @@
+
+Workflow histories are built based on event-sourcing, and each history event are persisted in Cadence storage. In Cadence, we call these history events <b>decision tasks</b>, the foundation of history replay. Most decision tasks have three status - <b>Scheduled</b>, <b>Started</b>, <b>Completed</b> and we will go over decision tasks produced by each Cadence operation in section below.
+
+When changing a workflow ownership of host and replaying a workflow, the decision tasks are downloaded from database and persisted in memory. Then during the workflow replaying process, if Cadence finds a decision task already exists for a particular step, it will immediately return the value of a decision task instead of rerunning the whole workflow logic. Let's take a look at the following simple workflow implementation and explictly list all decision tasks produced by this workflow.
+
+```go
+func SimpleWorkflow(ctx workflow.Context) error {
+	ao := workflow.ActivityOptions{
+		...
+	}
+	ctx = workflow.WithActivityOptions(ctx, ao)
+
+	var a int
+	err := workflow.ExecuteActivity(ctx, ActivityA).Get(ctx, &a)
+	if err != nil {
+		return err
+	}
+
+	workflow.Sleep(time.Minute)
+
+	err = workflow.ExecuteActivity(ctx, ActivityB, a).Get(ctx, nil)
+	if err != nil {
+		return err
+	}
+
+    workflow.Sleep(time.Hour)
+	return nil
+}
+```
+
+In this workflow, when it starts, it first execute ActivityA and then assign the result to an integer. It sleeps for one minute and then use the integer as an input argument to execute ActivityB. Finally it sleeps for one hour and completes.
+
+The following table lists the decision tasks stack produced by this workflow. It may looks overwhelming first but if you associate each decision task with its corresponding Cadence operation, it becomes self-explanatory.
+| ID | Decision Task Type | Explanation
+| -------- | ------- | -------------- |
+| 1 | Workflow Started
+| 2 | DecisionTaskScheduled  | workflow worker polling for work
+| 3 | DecisionTaskStarted    | worker gets the type `SimpleWorkflow`, lookup registred funcs, deserialize input, call it
+| 4 | DecisionTaskCompleted  | worker finishes
+| 5 | ActivityTaskScheduled  | activity worker polling for work
+| 6 | ActivityTaskStarted    | activity worker gets type `ActivityA` and do the job
+| 7 | ActivityTaskCompleted  | activity work completed with result of var a
+| 8 | DecisionTaskScheduled  | triggered by ActivityCompleted. server schedule next task
+| 9 | DecisionTaskStarted    |
+| 10 | DecisionTaskCompleted |
+| 11 | TimerStarted          | decision scheduled a timer for 1 minute
+| 12 | TimerFired            | fired after 1 minute
+| 13 | DecisionTaskScheduled | triggered by TimerFired
+| 14 | DecisionTaskStarted   |
+| 15 | DecisionTaskCompleted |
+| 16 | ActivityTaskScheduled | `ActivityB` scheduled by decision with param a
+| 17 | ActivityTaskStarted   | started by worker
+| 18 | ActivityTaskCompleted | completed with nil as error
```

Comment:
```suggestion
| 18 | ActivityTaskCompleted | completed with nil
```
"as error" feels kinda odd / implies something unique on this event because it's only on this one. 

might also be worth just leaving this entry empty since it's explained by the earlier one, and it's not relevant if it returned nil or not (the workflow doesn't check the return value, it could be anything or nothing).